lib/scholar/cluster/hierarchical.ex

```diff
@@ -0,0 +1,476 @@
+      clustering process.
+
+      Agglomerative clustering starts by considering each datum in `data` its own singleton group
+      or ["clade"](https://en.wikipedia.org/wiki/Clade).
+      It then picks two clades to merge into a new clade containing the data from both.
+      It does this until there is a single clade remaining.
+
+      Since each datum starts as its own clade, e.g. `data[0]` is clade `0`, indexing of new clades
+      starts at `n` where `n` is the size of the original `data` tensor.
+      If `clades[k] == [i, j]`, then clades `i` and `j` were merged to form `k + n`.
+
+    * `dissimilarities` (`Nx.Tensor` with shape `{n - 1}`) -
+      Contains a metric that measures the intra-clade closeness of each newly formed clade.
+      Represented by the heights of each clade in a dendrogram plot.
+      Determined by both the `:dissimilarity` and `:linkage` options.
+
+    * `num_points` (`pos_integer/0`) -
+      Number of points in the dataset.
+      Must be $\\geq 3$.
+
+    * `sizes` (`Nx.Tensor` with shape `{n - 1}`) -
+      `sizes[i]` is the size of clade `i`.
+      If clade `k` was created by merging clades `i` and `j`, then
+      `sizes[k] == sizes[i] + sizes[j]`.
+
+  ## Examples
+
+      iex> data = Nx.tensor([[2], [7], [9], [0], [3]])
+      iex> Hierarchical.fit(data)
+      %Scholar.Cluster.Hierarchical{
+        clades: Nx.tensor([[0, 4], [1, 2], [3, 5], [6, 7]]),
+        dissimilarities: Nx.tensor([1.0, 2.0, 2.0, 4.0]),
+        num_points: 5,
+        sizes: Nx.tensor([2, 2, 3, 5])
+      }
+  """
+  deftransform fit(%Nx.Tensor{} = data, opts \\ []) do
+    opts = NimbleOptions.validate!(opts, @fit_opts_schema)
+    dissimilarity = opts[:dissimilarity]
+    linkage = opts[:linkage]
+
+    dissimilarity_fun =
+      case dissimilarity do
+        # :precomputed -> &Function.identity/1
+        :euclidean -> &pairwise_euclidean/1
+      end
+
+    update_fun =
+      case linkage do
+        :average -> &average/6
+        # :centroid -> &centroid/6
+        :complete -> &complete/6
+        # :median -> &median/6
+        :single -> &single/6
+        :ward -> &ward/6
+        :weighted -> &weighted/6
+      end
+
+    dendrogram_fun =
+      case linkage do
+        # TODO: :centroid, :median
+        l when l in [:average, :complete, :single, :ward, :weighted] ->
+          &parallel_nearest_neighbor/2
+      end
+
+    n =
+      case Nx.shape(data) do
+        {n, _num_features} ->
+          n
+
+        other ->
+          raise ArgumentError,
+                "Expected a rank 2 (`{num_obs, num_features}`) tensor, found shape: #{inspect(other)}."
+      end
+
+    if n < 3 do
+      raise ArgumentError, "Must have a minimum of 3 data points, found: #{n}."
+    end
+
+    pairwise = dissimilarity_fun.(data)
+
+    if Nx.shape(pairwise) != {n, n} do
+      raise ArgumentError, "Pairwise must be a symmetric matrix."
+    end
+
+    {clades, diss, sizes} = dendrogram_fun.(pairwise, update_fun)
```

Comment:
Nitpick: we want to avoid going in and out of defn. So I would do this:

```suggestion
    {clades, diss, sizes} = dendrogram_fun.(data, dissimilarity_fun, update_fun)
```

The matrix shape check can be moved inside the `dendogram_fun`.

Reply:
So my hesitation on this is that I've got `dissimilarity: :precomputed` in mind. Note how Scipy supports that option:

https://docs.scipy.org/doc/scipy/reference/generated/scipy.cluster.hierarchy.linkage.html#scipy.cluster.hierarchy.linkage

> The input y may be either a 1-D condensed distance matrix or a 2-D array of observation vectors.

(Their "condensed distance matrix" is our pairwise.)

The idea is that building a dendrogram only requires pairwise, independent of what data was used to create it. If you have a custom dissimilarity fun, you can compute pairwise yourself and pass it in instead of the original data.

We can't do it yet, but once we support it we'll need to check pairwise independent of building the dendrogram.

Reply:
@billylanchantin then we can check for the shape on the dissimlarity: :precomputed branch or functions. In any case, the important is to call defn once.